framework/src/main/java/org/checkerframework/framework/util/FlowExpressionParseUtil.java

```diff
@@ -119,279 +82,156 @@ private static Pattern anchored(String regex) {
-            return parseParentheses(expression, context, path);
-        } else {
-            String message;
-            if (expression.equals("#0")) {
-                message =
-                        "one should use \"this\" for the receiver or \"#1\" for the first formal parameter";
+        if (expr != null) {
+            if (expr.isNullLiteralExpr()) {
+                return new ValueLiteral(types.getNullType(), (Object) null);
+            } else if (expr.isIntegerLiteralExpr() && !context.parsingMember) {
+                return new ValueLiteral(
+                        types.getPrimitiveType(TypeKind.INT), expr.asIntegerLiteralExpr().asInt());
+            } else if (expr.isLongLiteralExpr() && !context.parsingMember) {
+                return new ValueLiteral(
+                        types.getPrimitiveType(TypeKind.LONG), expr.asLongLiteralExpr().asLong());
+            } else if (expr.isDoubleLiteralExpr() && !context.parsingMember) {
+                return new ValueLiteral(
+                        types.getPrimitiveType(TypeKind.DOUBLE),
+                        expr.asDoubleLiteralExpr().asDouble());
+            } else if (expr.isStringLiteralExpr() && !context.parsingMember) {
+                TypeElement stringTypeElem =
+                        env.getElementUtils().getTypeElement("java.lang.String");
+                return new ValueLiteral(
+                        types.getDeclaredType(stringTypeElem),
+                        expr.asStringLiteralExpr().asString());
+            } else if (expr.isThisExpr()
+                    && !expr.asThisExpr().getTypeName().isPresent()
+                    && !context.parsingMember) {
+                return parseThis(context);
+            } else if (expr.isSuperExpr()) {
+                return parseSuper(types, context);
+            } else if (expr.isEnclosedExpr()) {
+                return parseHelper(expr.asEnclosedExpr().getInner(), context, path);
+            } else if (expr.isArrayAccessExpr()) {
+                return parseArray(expr.asArrayAccessExpr(), context, path);
+            } else if (expr.isNameExpr()
+                    && expr.asNameExpr().getNameAsString().startsWith("_param_")
+                    && !context.parsingMember) {
+                return parseParameter(expr.asNameExpr().getNameAsString(), context);
+            } else if (expr.isNameExpr()) {
+                return parseIdentifier(expr.asNameExpr().getNameAsString(), env, path, context);
+            } else if (expr.isMethodCallExpr()) {
+                return parseMethodCall(expr.asMethodCallExpr(), context, path, env);
+            } else if (expr.isFieldAccessExpr()) {
+                return parseMemberSelect(expr.asFieldAccessExpr(), env, context, path);
+            } else if (expr.isClassExpr() && !context.parsingMember) {
+                return parseHelper(
+                        StaticJavaParser.parseExpression(expr.asClassExpr().getTypeAsString()),
```

Comment:
Please add a comment here explaining why it is necessary to re-parse the type string.